---
title: >
  How to use `r Biocpkg("iSEE")` with HDF5
author:
- name: Kevin Rue-Albrecht
  affiliation: 
  - &id4 Kennedy Institute of Rheumatology, University of Oxford,
    Headington, Oxford OX3 7FY, UK.
  email: kevinrue67@gmail.com
- name: Federico Marini
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  - Center for Thrombosis and Hemostasis (CTH), Mainz
  email: marinif@uni-mainz.de
- name: Charlotte Soneson
  affiliation: 
  - &id3 Institute of Molecular Life Sciences, University of Zurich
  - SIB Swiss Institute of Bioinformatics
  email: charlottesoneson@gmail.com
- name: Aaron Lun
  affiliation: 
  - &id2 Cancer Research UK Cambridge Institute, University of Cambridge
  email: infinite.monkeys.with.keyboards@gmail.com 
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('iSEE')`"
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{6. Using iSEE with HDF5}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{iSEE}
  %\VignetteKeywords{GeneExpression, RNASeq, Sequencing, Visualization, QualityControl, GUI}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
bibliography: iSEE.bib
---

**Compiled date**: `r Sys.Date()`

**Last edited**: 2018-03-08

**License**: `r packageDescription("iSEE")[["License"]]`

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  error = FALSE,
  warning = FALSE,
  message = FALSE,
    crop = NULL
)
stopifnot(requireNamespace("htmltools"))
htmltools::tagList(rmarkdown::html_dependency_font_awesome())
sce <- readRDS("sce.rds")
```

```{r, eval=!exists("SCREENSHOT"), include=FALSE}
SCREENSHOT <- function(x, ...) knitr::include_graphics(x)
```

# Wrapping matrices in HDF5 files

Many `SummarizedExperiment` objects store assay matrices as in-memory matrix-like objects (e.g., ordinary `matrix`, `data.frame`, `DataFrame` object from the `r BiocStyle::Biocpkg("S4Vectors")` package, `sparseMatrix` derivative from the `r BiocStyle::CRANpkg("Matrix")` package.

```{r}
library(SingleCellExperiment)
class(assay(sce, "logcounts"))
```

In-memory matrices can become prohibitive for very large datasets and limited computational resources.
However, the HDF5 file format - and more generally on-disk objects - provides an alternative to work with large datasets.

> HDF5 is a unique technology suite that makes possible the management of extremely large and complex data collections.
> (Source: https://support.hdfgroup.org/HDF5/whatishdf5.html)

Specifically, wrapping an array-like object in a `DelayedArray` object from the `r BiocStyle::Biocpkg("DelayedArray")` package allows one to perform common array operations on it without loading the object in memory.
In order to reduce memory usage and optimize performance, operations on the object are either delayed or executed using a block processing mechanism.

Array-like objects such as matrices can be wrapped in an `HDF5Matrix` object as follows.
For demonstration purposes, we first make a copy of the original `sce` object to work with; we will be comparing the memory footprint of the two objects further below.

```{r}
sce_h5 <- sce
library(HDF5Array)
assay(sce_h5, "tophat_counts") <- writeHDF5Array(assay(sce_h5, "tophat_counts"))
class(assay(sce_h5, "tophat_counts"))
assay(sce_h5, "logcounts") <- writeHDF5Array(assay(sce_h5, "logcounts"))
class(assay(sce_h5, "logcounts"))
```

Comparing the memory footprint of the backed up original object `sce` and the HDF5-backed `sce_h5` object reveals the smaller memory footprint of `DelayedMatrix` objects.
Note that the memory footprint of `DelayedMatrix` objects is not just smaller - it is effectively constant size no matter the size of the dataset, as the full dataset only truly exists on disk, and read/write operations process only a small chunk of the dataset in memory at a time.

```{r}
format(object.size(sce), units = "Mb")
format(object.size(sce_h5), units = "Mb")
```

```{r, results='asis', echo=FALSE}
cat(sprintf(
  "In this case, using an HDF5 backend reduced the memory footprint %.1f-fold.",
  object.size(sce) / object.size(sce_h5)))
```

# iSEE supports HDF5-backed assays

We can then used the `SingleCellExperiment` object as usual.

```{r}
library(iSEE)
app <- iSEE(sce_h5)
```

Specifically, the use of the `SingleCellExperiment` class - itself derived from `SummarizedExperiment` enables `r BiocStyle::Biocpkg("iSEE")` to seamlessly support:

- any `r BiocStyle::Biocpkg("DelayedArray")` backend (e.g., `RleMatrix` objects; `ResidualMatrix`, `LowRankMatrix` from the `r BiocStyle::Biocpkg("BiocSingular")` package
- any matrix-like object in general that supports `dim()` and `[`

# Reduced dimensions

Dimensionality reduction results are also stored as matrices of coordinates, and can be similarly wrapped in `HDF5Matrix` objects.

Note that coercion - using the `as()` method, is a valid alternative that has the same effect as the `writeHDF5Array()` method.
However, an added benefit of `writeHDF5Array()` is the `filepath=` argument allowing users to control the HDF5 file where to write the dataset.

```{r}
sce_h5_rd <- sce_h5
for (i in seq_along(reducedDims(sce_h5_rd))) {
  reducedDim(sce_h5_rd, i) <- writeHDF5Array(reducedDim(sce_h5_rd, i))
  reducedDim(sce_h5_rd, i) <- as(reducedDim(sce_h5_rd, i), "HDF5Matrix")
}
```

As dimensionality reduction results already represent condensed representation of the dataset, the memory gain is smaller than for assays.

```{r}
format(object.size(sce_h5), units = "Mb")
format(object.size(sce_h5_rd), units = "Mb")
```

```{r, results='asis', echo=FALSE}
cat(sprintf(
  "Wrapping dimensionality reduction results further reduced the memory footprint %.1f-fold.",
  object.size(sce_h5) / object.size(sce_h5_rd)))
```

We can then used the `SingleCellExperiment` object as usual.

```{r}
app <- iSEE(sce_h5_rd)
```

# Session Info {.unnumbered}

```{r sessioninfo}
sessionInfo()
# devtools::session_info()
```

# References {.unnumbered}
