---
title: >
  How to use `r Biocpkg("iSEE")` with HDF5
author:
- name: Kevin Rue-Albrecht
  affiliation: 
  - &id4 Kennedy Institute of Rheumatology, University of Oxford,
    Headington, Oxford OX3 7FY, UK.
  email: kevinrue67@gmail.com
- name: Federico Marini
  affiliation: 
  - &id1 Institute of Medical Biostatistics, Epidemiology and Informatics (IMBEI), Mainz
  - Center for Thrombosis and Hemostasis (CTH), Mainz
  email: marinif@uni-mainz.de
- name: Charlotte Soneson
  affiliation: 
  - &id3 Institute of Molecular Life Sciences, University of Zurich
  - SIB Swiss Institute of Bioinformatics
  email: charlottesoneson@gmail.com
- name: Aaron Lun
  affiliation: 
  - &id2 Cancer Research UK Cambridge Institute, University of Cambridge
  email: infinite.monkeys.with.keyboards@gmail.com 
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('iSEE')`"
output: 
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{3. Using with HDF5}
  %\VignetteEncoding{UTF-8}  
  %\VignettePackage{iSEE}
  %\VignetteKeywords{GeneExpression, RNASeq, Sequencing, Visualization, QualityControl, GUI}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
bibliography: iSEE.bib
---

**Compiled date**: `r Sys.Date()`

**Last edited**: 2018-03-08

**License**: `r packageDescription("iSEE")[["License"]]`

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  error = FALSE,
  warning = FALSE,
  message = FALSE,
    crop = NULL
)
stopifnot(requireNamespace("htmltools"))
htmltools::tagList(rmarkdown::html_dependency_font_awesome())
sce <- readRDS("sce.rds")
```

```{r, eval=!exists("SCREENSHOT"), include=FALSE}
SCREENSHOT <- function(x, ...) knitr::include_graphics(x)
```

# Wrapping matrices in HDF5 files

Let us make a copy of the originak `sce` object to work with; we will be comparing the memory footprint of the object using an HDF5 backend.

```{r}
sce_h5 <- sce
```

Many `SummarizedExperiment` objects store assay matrices as regular `matrix` objects.

```{r}
library(SingleCellExperiment)
class(assay(sce_h5, "logcounts"))
```

This type of base `matrix` is fully loaded in memory, which can become prohibitive for very large datasets.

However, the HDF5 file format provides an alternative to work with large datasets.

> HDF5 is a unique technology suite that makes possible the management of extremely large and complex data collections.
> (Source: https://support.hdfgroup.org/HDF5/whatishdf5.html)

Array-like objects such as matrices can be wrapped in an `HDF5Matrix` object as follows.

```{r}
library(HDF5Array)
assay(sce_h5, "tophat_counts") <- writeHDF5Array(assay(sce_h5, "tophat_counts"))
class(assay(sce_h5, "tophat_counts"))
assay(sce_h5, "logcounts") <- writeHDF5Array(assay(sce_h5, "logcounts"))
class(assay(sce_h5, "logcounts"))
```

Comparing the memory footprint of the backed up original object `sce` and the HDF5-backed `sce_h5` object reveals the smallermemory footprint of HDF5-backed objects.

```{r}
format(object.size(sce), units = "Mb")
format(object.size(sce_h5), units = "Mb")
```

```{r, results='asis', echo=FALSE}
cat(sprintf(
  "In this case, using an HDF5 backend reduced the memory footprint %.1f-fold.",
  object.size(sce) / object.size(sce_h5)))
```

# iSEE supports HDF5-backed assays

We can then used the `SingleCellExperiment` object as usual.

```{r}
library(iSEE)
app <- iSEE(sce_h5)
```

# Reduced dimensions

Dimensionality reduction results are also stored as matrices of coordinates, and can be similarly wrapped in `HDF5Matrix` objects.

Note that coercion - using the `as()` method, is a valid alternative that has the same effect as the `writeHDF5Array()` method.
However, an added benefit of `writeHDF5Array()` is the `filepath=` argument allowing users to control the HDF5 file where to write the dataset.

```{r}
sce_h5_rd <- sce_h5
for (i in seq_along(reducedDims(sce_h5_rd))) {
  reducedDim(sce_h5_rd, i) <- writeHDF5Array(reducedDim(sce_h5_rd, i))
  reducedDim(sce_h5_rd, i) <- as(reducedDim(sce_h5_rd, i), "HDF5Matrix")
}
```

As dimensionality reduction results already represent condensed representation of the dataset, the memory gain is smaller than for assays.

```{r}
format(object.size(sce_h5), units = "Mb")
format(object.size(sce_h5_rd), units = "Mb")
```

```{r, results='asis', echo=FALSE}
cat(sprintf(
  "Wrapping dimensionality reduction results further reduced the memory footprint %.1f-fold.",
  object.size(sce_h5) / object.size(sce_h5_rd)))
```

We can then used the `SingleCellExperiment` object as usual.

```{r}
app <- iSEE(sce_h5_rd)
```

# Session Info {.unnumbered}

```{r sessioninfo}
sessionInfo()
# devtools::session_info()
```

# References {.unnumbered}
